### PR TITLE
Issue 77 send only modified changes

### PR DESCRIPTION
**Pull request by Derek52**
Here is what I have so far. I think it should work. I unit tested the javascript separately(I can share my tests with you), so I know that is good.

I was unable to figure out how to test the kotlin function though. I spent a little time trying to figure out how to create an element in a test and play with changing it's value, and I didn't get too far. If you want to point me towards how to do that, I can work on that though.

### CODE REVIEW

File: src/main/kotlin/kweb/prelude.kt [hunk 1]
```diff
@@ -434,16 +437,38 @@ abstract class ValueElement(open val element: Element, val kvarUpdateEvent: Stri
             _valueKvar = v
         }
 
+
     /**
      * Automatically update `toBind` with the value of this INPUT element when `updateOn` event occurs.
      */
+
+    //I wasn't sure where this data class and function should go
+    @Serializable
```
Comment (sanity) [inline src/main/kotlin/kweb/prelude.kt:446]: I think it makes sense to keep them near where they are used.

File: src/main/kotlin/kweb/prelude.kt [hunk 2]
```diff
@@ -434,16 +437,38 @@ abstract class ValueElement(open val element: Element, val kvarUpdateEvent: Stri
...
         setValue(toBind as KVal<String>)
 
         // TODO: Would be really nice if it just did a diff on the value and sent that, rather than the
         //       entire value each time PARTICULARLY for large inputs
-        on(retrieveJs = "${jsExpression}.value").event(updateOn, Event::class) {
-            toBind.value = it.retrieved ?: error("No value was retrieved")
+        on(retrieveJs = "get_diff_changes(${jsExpression})").event(updateOn, Event::class) {
+            //I think this line could be made better to look at. But I think it works.
+            //I hate having to use the !!, it'd be better to catch the null and do some error handling
+            //But, I don't know really know what this function should do if it's sent a bad input
```
Comment (sanity) [inline src/main/kotlin/kweb/prelude.kt:469]: It should probably throw an exception if the input is unexpected, similar to what was there before `it.retrieved ?: error("No value was retrieved")`

File: src/main/kotlin/kweb/util/Template.kt [hunk 3]
```diff
@@ -1,4 +1,4 @@
-package kweb.common
+package kweb.util
```
Comment (sanity) [inline src/main/kotlin/kweb/util/Template.kt:1]: Was there a reason for the package name change? Not opposed to it, just curious.

File: src/main/resources/kweb/kweb_bootstrap.js [hunk 4]
```diff
@@ -203,6 +203,62 @@ function removeClass(el, className) {
...
+    constructor(prefixEndIndex, postfixOffset, diffString) {
+        this.prefixEndIndex = prefixEndIndex;
+        this.postfixOffset = postfixOffset;
+        this.diffString = diffString;
+    }
+}
+
+function get_diff_changes(id) {
+    let e = document.getElementById(id);
+    let newString = e.dataset.value;//reads the new string value from data-attribute data-vale
```
Comment (sanity) [inline src/main/resources/kweb/kweb_bootstrap.js:216]: Nice, I wasn't aware of `dataset`.

File: src/main/resources/kweb/kweb_bootstrap.js [hunk 5]
```diff
@@ -203,6 +203,62 @@ function removeClass(el, className) {
...
+    let oldString = e.dataset.previousInput;//reads the oldString from the data-attribute data-previous-input
+
+    //This seems like a logical place to do this, but I'm unsure if it is safe to do so
+    //I think it might prematurely change the value of toBind.value in prelude.kt's setValue function
+    //savePreviousInput(newString, id)//put the newString into the data attribute so it can be used as the oldString the next time this method is run
+
+    if (oldString === undefined) {//the first time this is run previous-input should be undefined so we just return the new string
+        return new DiffPatchData(0, 0, newString);
+    }
+    let oldArray = Array.from(oldString);
```
Comment (sanity) [inline src/main/resources/kweb/kweb_bootstrap.js:226]: Why the conversion to Arrays - can the strings not be iterated over directly, eg. using `str.charAt()`?

File: src/main/resources/kweb/kweb_bootstrap.js [hunk 6]
```diff
@@ -203,6 +203,62 @@ function removeClass(el, className) {
...
+        if (oldArray[i] == newArray[i]) {
+            commonPrefixEnd = i+1;
+        } else break;
+    }
+    for(let offset = 0; offset < shorterStringLength - commonPrefixEnd; offset++) {
+        if (oldArray[oldStringLastIndex - offset] == newArray[newStringLastIndex - offset]) {
+            commonPostfixOffset = offset;
+        } else break;
+    }
+    if (oldStringLastIndex > newStringLastIndex) {//if oldString is longer we are removing text and we send an empty string as the diff
```
Comment (sanity) [inline src/main/resources/kweb/kweb_bootstrap.js:248]: Wondering if this special case is necessary, what happens if we don't check for this condition?

File: src/main/resources/kweb/kweb_bootstrap.js [hunk 7]
```diff
@@ -203,6 +203,62 @@ function removeClass(el, className) {
...
+        return new DiffPatchData(commonPrefixEnd, commonPostfixOffset, "");
+    } else {
+        return new DiffPatchData(commonPrefixEnd, commonPostfixOffset,
+            newString.substring(commonPrefixEnd, newString.length - commonPostfixOffset));
+    }
+}
+
+
+function savePreviousInput(previousInputString, id) {
```
Comment (sanity) [inline src/main/resources/kweb/kweb_bootstrap.js:257]: Looks like this is unused now.

**Review by sanity — CHANGES_REQUESTED** [review]

File: src/main/kotlin/kweb/util/Template.kt (hunk 3 above)
Comment (Derek52) [reply src/main/kotlin/kweb/util/Template.kt:1]: I just randomly found it as I was clicking through the code. IntelliJ had it labeled as a warning, and I thought it made sense to change the package name to the package the file was actually in. It's more OCD than anything.

File: src/main/resources/kweb/kweb_bootstrap.js (hunk 6 above)
Comment (Derek52) [reply src/main/resources/kweb/kweb_bootstrap.js:248]: I just tested it, and it turns out it isn't necessary to test for this case. I'll change it.

File: src/main/resources/kweb/kweb_bootstrap.js (hunk 5 above)
Comment (Derek52) [reply src/main/resources/kweb/kweb_bootstrap.js:226]: I just prefer working with arrays. I made a mental note to convert it to using str.charAt() after I had the logic fleshed out, and kind of forgot to do that. It should be a quick fix for me to go and change it though.

File: src/main/resources/kweb/kweb_bootstrap.js (hunk 7 above)
Comment (Derek52) [reply src/main/resources/kweb/kweb_bootstrap.js:257]: I put that there for use in the get_diff_changes() function. I commented out the part where I call it though, because I wasn't 100% sure it was safe to do so though. I was worried about prematurely changing the data-attribute that stored the previous input.

File: src/main/kotlin/kweb/util/Template.kt (hunk 3 above)
Comment (sanity) [reply src/main/kotlin/kweb/util/Template.kt:1]: Ah, didn't realize the package didn't match the dir.

**Comment by sanity** [comment]
I wrote this in Whatsapp but will put it here to keep everything in the same place:

[9:04 AM, 2/3/2021] Ian Clarke: I just responded to your pull request, sorry for the delay. Unit testing browser-server interactions like this one is a little more involved, you can see an example in TodoDemoTest, it uses this unit testing tool to do it: https://bonigarcia.github.io/selenium-jupiter/
[9:05 AM, 2/3/2021] Ian Clarke: If you're running the code locally you should be able to use all the normal Kotlin debugging tools in Kotlin code, and you can use JavaScript debugging tools in the browser.
[9:06 AM, 2/3/2021] Ian Clarke: Kweb's unit testing isn't super-thorough right now, but I'd like to try to improve it over time so it might be good practice to get accustomed to unit testing now.
[9:07 AM, 2/3/2021] Ian Clarke: Only thing is that selenium-jupiter can be a bit slow as it actually launches a web browser from within the unit test, but we can optimize that later

**Review by sanity — COMMENTED** [review]
These look good, once they're reasonably well tested I think we'll be good to merge.

**Comment by Derek52** [comment]
I've added the tests. I just put them in a file called PreludeTest.kt. I wanted to give the file a name specific to this issue, but, couldn't think of one that wasn't way too long.

**Comment by sanity** [comment]
Looks like the unit tests aren't working yet: https://github.com/kwebio/kweb-core/pull/176/checks?check_run_id=1846612945

**Comment by Derek52** [comment]
I bumped up the version numbers for kotlin coroutines, kotlin-jvm, and kotlin-serialization, and it seems to have removed the issue locally. We'll see if it works here.

**Comment by Derek52** [comment]
1 of the tests failed, the other 5 passed, and the error log is much shorter. The test that failed passes locally, almost every time. For some reason one of my tests has failed a few times I've run the suite, but passes every other time I run the tests. 

I think this intermittent failing is being caused by the way I wrote the tests. I'm using a lateinit, and I think that variable is, sometimes, being accessed before it's initialized. Or, maybe there is a race condition somewhere in the code. I'm gonna have to spend more time investigating.

**Comment by sanity** [comment]
Is this the failure?

```
PreludeTest > removeTextFromMiddle(WebDriver) FAILED
    org.opentest4j.AssertionFailedError at PreludeTest.kt:90
```

That looks like it's this line:

```
        inputField.getAttribute("value").shouldBe("Lazy Fox")
```

Is that what you're seeing?

**Comment by Derek52** [comment]
After bumping up the version numbers, yes, that's what I'm seeing, in the github logs. Not, locally. Locally, that assertion passes. Mostly. I have seen it fail locally a couple of times. But, then I will run it again, without changing any code, and it will pass. It fails intermittently.

**Comment by sanity** [comment]
Do you know what's being returned by `inputField.getAttribute("value")` when it fails?

**Comment by Derek52** [comment]
I had to run all my tests 8 times to get one to fail again. I'll paste the error log. "Lazy Brown Fox", the default string, was returned. So, the method where I changed it didn't run. Or it ran in a thread and wasn't done when the value was read. Looking at the error log, I'm thinking this is a problem with selenium. I'm going to write my tests in a different way and see if this fixes the issue.

expected: "Brown Fox" but was: "Lazy Brown Fox"
org.opentest4j.AssertionFailedError: expected: "Brown Fox" but was: "Lazy Brown Fox"
	at kweb.PreludeTest.removeTextFromBeginning(PreludeTest.kt:82)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:688)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:140)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:84)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$6(TestMethodTestDescriptor.java:210)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:206)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:131)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:65)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
	at java.util.ArrayList.forEach(ArrayList.java:1259)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:143)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
	at java.util.ArrayList.forEach(ArrayList.java:1259)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:143)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:32)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:51)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:108)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:96)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:75)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.processAllTestClasses(JUnitPlatformTestClassProcessor.java:99)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.access$000(JUnitPlatformTestClassProcessor.java:79)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor.stop(JUnitPlatformTestClassProcessor.java:75)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:61)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy2.stop(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.stop(TestWorker.java:133)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:414)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
	at java.lang.Thread.run(Thread.java:748)

**Comment by Derek52** [comment]
After trying to rewrite the tests and create the WebDriver a few different ways. I learned that a simple Thread.sleep(300), makes all the tests pass every time. WebDriver is not thread safe, and timing issues do happen.

Now I know that my logic is good, and that WebDriver has thread issues, I'll try to rewrite the tests in a way that won't require the sleep hack to pass.

**Comment by sanity** [comment]
Yeah, I don't love the browser testing framework, I've had a few fights with it myself but it was the least bad option I could find.

**Comment by Derek52** [comment]
I tried breaking the tests out into multiple files to see if that would help. I tried a few other things today too. I have yet to find a way that can make these tests consistently pass that doesn't use a Thread.sleep.

I've also discovered that my get_changes function is called on every new key press in an input field. I believe you said you wanted to avoid that, but I'm not quite sure how to change that behavior at the moment. I found this out by writing a test that used sendKeys(), and using a non-headless browser. I manually checked the javascript console in Chrome as it ran.


    fun setValue(toBind: KVar<String>, updateOn: String = "input") {
        setValue(toBind as KVal<String>)

        // TODO: Would be really nice if it just did a diff on the value and sent that, rather than the
        //       entire value each time PARTICULARLY for large inputs
        on(retrieveJs = "get_diff_changes(${jsExpression})").event(updateOn, Event::class) {
            val diffDataJson = it.retrieved ?: error("No diff data was retrieved")
            val diffData = Json.decodeFromString<DiffData>(diffDataJson)
            toBind.value = applyDiff(toBind.value, diffData)
        }
    }

Also, I have tried running my tests with the debugger, and for some reason, in the block above, diffData is shown as null. I can't wrap my head around why. Using my browser's console I can see that get_diff_changes() runs, and I used a console.log to make sure it was getting the right result. But, I can't see an actual value show up in the Kotlin code.

I tried to write some tests that would let me clearly check on the state of this variable(and a couple of other ones), but I couldn't figure out how to interface with either Javascript code or data, in a test. I also struggled with Selenium WebElement vs KWeb Element.
